devices/flash-zynq7000/flashcfg.c

```diff
@@ -46,128 +47,270 @@
-	info->cmds[flash_cmd_4qpp].dummyCyc = 0;
+	info->init = flashcfg_spansionInit;
+}
 
-	info->cmds[flash_cmd_p4e].opCode = FLASH_CMD_P4E;
-	info->cmds[flash_cmd_p4e].size = 4;
-	info->cmds[flash_cmd_p4e].dummyCyc = 0;
 
-	info->cmds[flash_cmd_4p4e].opCode = FLASH_CMD_4P4E;
-	info->cmds[flash_cmd_4p4e].size = 5;
-	info->cmds[flash_cmd_4p4e].dummyCyc = 0;
+static int flashcfg_micronInitVolatileCRWriteEnable(const flash_info_t *info, int enable)
+{
+	int res;
+	u8 tx[3];
+
+	res = flashcfg_wren(info);
+	if (res < 0) {
+		return res;
+	}
+
+	/* Enable non-volatile CR write */
+	hal_memset(tx, 0, sizeof(tx));
+	/* non-volatile CR write cmd code */
+	tx[0] = 0xb1;
+
+	/* Output driver strength 30 Ohms, TODO optimize */
+	/* Enable hold, disable Quad I/O and Dual I/O protocols */
+	/* Enable write on non-volatile register */
+	tx[1] = 0xfe;
+	if (enable == 0) {
+		tx[1] |= 1;
+	}
+
+	/* Set default value for number of dummy clock cycles */
+	/* Disable XIP */
+	tx[2] = 0xff;
 
-	info->cmds[flash_cmd_p64e].opCode = FLASH_CMD_SE;
-	info->cmds[flash_cmd_p64e].size = 4;
-	info->cmds[flash_cmd_p64e].dummyCyc = 0;
+	qspi_start();
+	res = qspi_polledTransfer(tx, NULL, 3, FLASH_TIMEOUT_CMD_MS);
+	qspi_stop();
 
-	info->cmds[flash_cmd_4p64e].opCode = FLASH_CMD_4SE;
-	info->cmds[flash_cmd_4p64e].size = 5;
-	info->cmds[flash_cmd_4p64e].dummyCyc = 0;
+	if (res < 0) {
+		return res;
+	}
 
-	info->cmds[flash_cmd_be].opCode = FLASH_CMD_BE;
-	info->cmds[flash_cmd_be].size = 1;
-	info->cmds[flash_cmd_be].dummyCyc = 0;
+	return flashcfg_wipCheck(info, FLASH_TIMEOUT_WIP_MS);
 }
 
 
-void flashcfg_jedecIDGet(flash_cmd_t *cmd)
+static int flashcfg_micronInitVolatileCRSetDummyCyc(const flash_info_t *info)
 {
-	cmd->opCode = FLASH_CMD_RDID;
-	cmd->size = 1;
-	cmd->dummyCyc = 24;
+	int res;
+	u8 tx[2];
+
+	if ((info->cmds[info->readCmd].dummyCyc - 1) > 0xf) {
+		return -EINVAL;
+	}
+
+	/* Volatile CR write cmd code */
+	tx[0] = 0x81;
+
+	/* Set number of dummy cycles */
+	tx[1] = ((info->cmds[info->readCmd].dummyCyc - 1) << 4);
+	/* Disable XIP, continuous read */
+	tx[1] |= 0xb;
+
+	qspi_start();
+	res = qspi_polledTransfer(tx, NULL, 2, FLASH_TIMEOUT_CMD_MS);
+	qspi_stop();
+
+	if (res < 0) {
+		return res;
+	}
+
+	return flashcfg_wipCheck(info, FLASH_TIMEOUT_WIP_MS);
 }
 
 
-static void flashcfg_spansion(flash_info_t *info)
+static int flashcfg_micronInit(const flash_info_t *info)
 {
-	info->name = "Spansion s25fl256s1";
-	flashcfg_defaultCmds(info);
+	int res;
 
-	/* FIXME: s25fl256s1 should be able to operate in quad and dual mode for 4byte transactions.
-	          In the current implementation errors occur for the dual and the quad mode. */
-	info->readCmd = flash_cmd_4fast_read;
-	info->ppCmd = flash_cmd_4pp;
+	res = flashcfg_micronInitVolatileCRWriteEnable(info, 1);
+	if (res < 0) {
+		return res;
+	}
+	res = flashcfg_micronInitVolatileCRSetDummyCyc(info);
+	if (res < 0) {
+		return res;
+	}
+	res = flashcfg_micronInitVolatileCRWriteEnable(info, 0);
```

Comment:
All these functions are used exactly once and have java-script-like names - how about implementing them inside of `flashcfg_micronInit`?